### PR TITLE
Fix URN response handling

### PR DESCRIPTION
**Pull request by eduard-bagdasaryan**
urnHandleReply() may be called several times while copying the entry
from the store. Each time it must use the buffer length that is left
(from the previous call).

Also do not abandon a urn entry, still having clients attached.

Also allow urnHandleReply() to produce a reply if it receives a
zero-sized buffer. This may happen after the entry has been fully
stored.

### CODE REVIEW

File: src/urn.cc [hunk 1]
```diff
@@ -82,6 +82,11 @@ CBDATA_CLASS_INIT(UrnState);
 
 UrnState::~UrnState()
 {
+    if (urlres_e) {
+        storeUnregister(sc, urlres_e, this);
+        urlres_e->unlock("~UrnState+res");
+        entry->unlock("~UrnState+prime");
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/urn.cc]: Done at 7b309bd.

**Comment by eduard-bagdasaryan** [comment]
Please mark with S-waiting-for-reviewer label.

**Comment by eduard-bagdasaryan** [comment]
@yadij , have you approved this PR? I see that you requested changes, but I have not found them.

**Comment by yadij** [comment]
Sorry, I left it on comment instead of approve.

**Comment by rousskov** [comment]
We will fix the destructors-must-not-throw violation introduced by this PR.

**Pull request closed without merging**